util/fetch/src/client.rs

```diff
@@ -14,347 +14,554 @@
+		if self.refs.fetch_sub(1, Ordering::SeqCst) == 1 {
+			// ignore send error as it means the background thread is gone already
+			let _ = self.core.clone().send(None).wait();
+		}
 	}
+}
+
+impl Client {
+	/// Create a new fetch client.
+	pub fn new() -> Result<Self, Error> {
+		let startup_done = Arc::new((Mutex::new(Ok(())), Condvar::new()));
+		let (tx_proto, rx_proto) = mpsc::channel(64);
+
+		Client::background_thread(startup_done.clone(), rx_proto)?;
+
+		let mut guard = startup_done.0.lock();
+		let startup_result = startup_done.1.wait_for(&mut guard, Duration::from_secs(3));
+
+		if startup_result.timed_out() {
+			error!(target: "fetch", "timeout starting background thread");
+			return Err(Error::BackgroundThreadDead)
+		}
+		if let Err(e) = mem::replace(&mut *guard, Ok(())) {
+			error!(target: "fetch", "error starting background thread: {}", e);
+			return Err(e.into())
+		}
 
-	fn with_limit(limit: Option<usize>) -> Result<Self, Error> {
 		Ok(Client {
-			client: RwLock::new((time::Instant::now(), Self::new_client()?)),
-			pool: CpuPool::new(4),
-			limit: limit,
+			core: tx_proto,
+			refs: Arc::new(AtomicUsize::new(1)),
 		})
 	}
 
-	/// Sets a limit on the maximum download size.
-	pub fn set_limit(&mut self, limit: Option<usize>) {
-		self.limit = limit
+	fn background_thread(start: StartupCond, rx_proto: mpsc::Receiver<ChanItem>) -> io::Result<thread::JoinHandle<()>> {
+		thread::Builder::new().name("fetch".into()).spawn(move || {
+			let mut core = match reactor::Core::new() {
+				Ok(c)  => c,
+				Err(e) => {
+					*start.0.lock() = Err(e);
+					start.1.notify_one();
+					return ()
+				}
+			};
+			let handle = core.handle();
+			let hyper = hyper::Client::configure()
+				.connector(hyper_rustls::HttpsConnector::new(4, &core.handle()))
+				.build(&core.handle());
+
+			let future = rx_proto.take_while(|item| Ok(item.is_some()))
+				.map(|item| item.expect("`take_while` is only passing on channel items != None"))
+				.for_each(|(url, abort, sender)|
+			{
+				trace!(target: "fetch", "new request to {}", url);
+				if abort.is_aborted() {
+					return future::ok(sender.send(Err(Error::Aborted)).unwrap_or(()))
+				}
+				let ini = (hyper.clone(), url, abort, 0);
+				let fut = future::loop_fn(ini, |(client, url, abort, redirects)| {
+					let url2 = url.clone();
+					let abort2 = abort.clone();
+					client.request(get(&url))
+						.map(move |resp| Response::new(url2, resp, abort2))
+						.from_err()
+						.and_then(move |resp| {
+							if abort.is_aborted() {
+								debug!(target: "fetch", "fetch of {} aborted", url);
+								return Err(Error::Aborted)
+							}
+							if let Some(next_url) = redirect_location(url, &resp) {
+								if redirects >= abort.max_redirects() {
+									return Err(Error::TooManyRedirects)
+								}
+								Ok(Loop::Continue((client, next_url, abort, redirects + 1)))
+							} else {
+								let content_len = resp.headers.get::<ContentLength>().cloned();
```

Comment:
What about chunked responses? Seems that we should handle them in `BodyReader`?

Reply:
Right, I updated `BodyReader` to track the number of bytes read and error out if the maximum is exceeded. While this logic was part of the `Stream` impl of `Response` it wasn't in `BodyReader`.